Реализовано обращение к телеграм

GlazProject opened the pull request:
Получение данных бота из конфигурационного файла. Разделение на обработчики команд и обращение к телеграм

GlazProject committed:
Реализовано обращение к телеграм. Получение данных бота из конфигурационного файла. Разделение на обработчики команд и обращение к телеграм

GlazProject committed:
Fix:
1) обновлена логика работы обработчика команд. Команды как список
2) класс TelegramBot реализует Writer
3) классу TelegramBot добавлен читатель, который преобразует обновление в RequestData
4) Все обработчики переведены на Request и Response data. Каждый обработчик сам проверяет соответчитвие версии контейнеров для обработки
5) Добавлены версии к обёрткам данных. V1 содержит только поле со строкой.
6) Исправлены простые замечания
7) Добавлены зависимости логирования

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/Main.java:
```diff
@@ -0,0 +1,72 @@
+
+    public static void telegramEchoBot() throws FileNotFoundException {
+        Properties properties = new Properties();
+        try {
+            properties.load(new FileInputStream("src/main/resources/config.properties"));
+        } catch (IOException e) {
+            throw new FileNotFoundException("Файл конфигурации не найден по адресу ./resources/config.properties");
+        }
```
На самом деле, тут вполне можно прокидывать сразу наверх это исключение, потому что в итоговой ошибке будет примерно тоже самое :) но такой подход тоже вполне норм) Можно даже в свою специальную ошибку заворачивать

GlazProject replied:
В каком смысле пробрасывать наверх? Разве throw не передаёт её вышестоящему уровню? Как задать запрос, чтобы почитать об этом?

TheDoctorTimeLord replied:
Есть оператор `throws`. В данном случае выглядело бы это так:
```
public static void telegramBot() throws IOException {
        properties.load(new FileInputStream("src/main/resources/config.properties"));
        ...
}
```

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/Main.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+
+    public static void telegramEchoBot() throws FileNotFoundException {
```
Думаю, что можно метод сразу переименовать в `telegramBot()`

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/Handler.java:
```diff
@@ -1,16 +1,13 @@
      * @param data Данные полученные при чтении
      * @return Контейнер с обработанными данными
      */
-    ResponseData handle(RequestData data);
```
Предлагаю оставить ResponseData и RequestData. Лучше уметь сообщения телеграма мапить в какое-то необходимое нам представление. Тогда логика будет универсальна для любого бота, а не только для телеграма, при этом мы сохраним преимущество объекта перед строками

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/CommandsHandler.java:
```diff
@@ -0,0 +1,56 @@
+        public String text = "";
+    }
+
+    private final Map<String, Handler<String,ResponseData>> handlers;
```
Предлагаю изменить структуру. Давай для обработки команд мы будем хранить `List<Handler<RequestData, ResponseData>>` (здесь можно у `Handler` уйти от Generic к тому, что было).
1. Изначально мы в этот список сохраняем обработчики сообщения в том порядке, в котором мы бы хотели пытаться обрабатывать текст (например, сначала команды, потом текст).
2. Для каждого `RequestData` от пользователя мы итерируемся по нашему списку, передавая в обработчик  исходный запрос.  Если очередной обработчик в результате вызова `handle` вернул нам объект `ResponseDate`, то мы считаем, что обработка запроса завершена и возвращаем ответ. Если вместо него мы получили `null` - то значит данные обработчик не может обработать такое сообщение, и мы переходим к следующему

Такой вариант нам поможет:
1. Избавиться от необходимости редактирования логики этого класса (только добавлять обработчики)
2. Избавиться от необходимости парсить ответ каким-то универсальным образом (здесь это делает `findCommand`)
3. Избавиться от необходимости создавать `Command`
4. Добиться лёгкой расширяемости за счёт создания новых обработчиков, а не изменения старых

Также могу заметить, что в такой структуре мы можем использовать паттерн `Facade`, для того, чтобы упростить работу с обработчиками команд (не текста). Мы можем объединить все команды в общем классе `CommandsFacade`. Этот класс в методе `handle` будет иметь логику подбора команды. В свою очередь команды можно создавать и хранить в виде `Map<String, Handler>`, где ключом выступает название команды, как это сделано здесь.

Если возникнут вопросы - пишите в ЛС

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/CommandsHandler.java:
```diff
@@ -0,0 +1,56 @@
+        return new UnknownCommandHandler().handle(command.text);
+    }
+
+    private void fillHandlers(){
+        handlers.put("", new TextHandler());
+        handlers.put("start", new StartCommandHandler());
+    }
```
В Java есть альтернативный способ заполнения Map константным наполнением:
```
ImmutableMap.builder()
    .put("", new TextHandler())
    .put("start", new StartCommandHandler())
    .build();
```
Это применение шаблона программирования `Строитель` или `Builder`. Преимущество в данном случае в том, что вы можете делать инициализацию прям сразу в поле класса, без необходимости создавать отдельный метод

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/CommandsHandler.java:
```diff
@@ -0,0 +1,56 @@
+        handlers.put("start", new StartCommandHandler());
+    }
+
+    private Command findCommand(String inputText){
```
Предлагаю по итогам написанного выше избавиться от этого метода :)

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/textHandlers/TextHandler.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public ResponseData handle(String data) {
+        return new SimpleEchoHandler().handle(data);
```
Вынесем создание `SimpleEchoHandler` в константы или поле класса?

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/readers/RequestData.java:
```diff
@@ -1,24 +1,24 @@
  */
-public class RequestData {
-    private String dataText;
+public class RequestData<T> {
```
Не думаю, что тут есть такая необходимость Generic... Только если планируется в дальнейшем в этом поле хранить какие-то сложные данные, при том, что сам `RequestData` будет хранить данные о пользователе... Пока что не уверен в полезности такой конструкции

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/Bot.java:
```diff
@@ -0,0 +1,42 @@
+import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
+
+public class Bot extends TelegramLongPollingBot {
+    private final String BOT_NAME;
+    private final String BOT_TOKEN;
```
В данном случае это просто поля объекта, поэтому имена нужно писать camel case.

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/Bot.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void onUpdateReceived(Update update) {
+        TelegramHandler handler = new TelegramHandler();
+        TelegramWriter writer = new TelegramWriter();
```
Давай создадим это в конструкторе класса

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/TelegramWriter.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
+
+public class TelegramWriter {
+    public SendMessage createMessage(TelegramResponseData responseData){
+        SendMessage message = new SendMessage();
+        message.setChatId(responseData.getMessage().getChatId());
+        message.setText(responseData.getText());
+        if (message.getText().equals(""))
+            message.setText("\uE000");
+        return message;
+    }
```
Наша изначальная идея заключалась в том, что мы избавляемся от частностей в пользу абстракций. Здесь правильнее всего было бы взять наш изначальный интерфейс `Writer`, имплементировать его в классе нашего бота, реализовать в методе `void write(ResponseData)` отправку сообщения в телеграм. Теперь, если где-то нам будет необходим способ отправить сообщение пользователю, мы сможем передать туда наш `Bot` в виде `Writer`. Как это применить для нашего способа обработки - смотри дальше

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/Bot.java:
```diff
@@ -0,0 +1,42 @@
+        TelegramHandler handler = new TelegramHandler();
+        TelegramWriter writer = new TelegramWriter();
+        TelegramResponseData telegramRequestData = handler.handle(update);
+        SendMessage message = writer.createMessage(telegramRequestData);
+        try{
+            execute(message);
+        }
+        catch (TelegramApiException e){
+            e.printStackTrace();
+        }
```
Если мы реализуем тут интерфейс `Writer`, то весь этот код замениться на `write(telegramRequestData )`. Соответственно, тип переменной `telegramRequestData` должен стать `ResponseData`

GlazProject replied:
Мы не можем отправлять сообщение ботом, если не сформируем его. А сформировать сообщение мы можем только из принятого сообщения, откуда возьмём id чата. Поэтому создавать сообщение всё равно придётся из `TelegramRequestData`

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/Bot.java:
```diff
@@ -0,0 +1,42 @@
+    public void onUpdateReceived(Update update) {
+        TelegramHandler handler = new TelegramHandler();
+        TelegramWriter writer = new TelegramWriter();
+        TelegramResponseData telegramRequestData = handler.handle(update);
```
`telegramRequestData` -> `telegramResponseData`

GlazProject committed:
Fix:
1) Переведены классы данных на компоненты-делегаты
2) Реализован фасад для обработчиков

GlazProject committed:
Fix:
1) Переведены классы данных на компоненты-делегаты
2) Реализован фасад для обработчиков

GlazProject committed:
Добавлены описания ко всем методам и классам
Переписана работа обработчика Telegram
Добавлен класс тестирования

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/Main.java:
```diff
@@ -0,0 +1,78 @@
+
+    public static void main(String[] args) throws IOException {
+//        telegramBot();
+        consoleEchoBot();
```
Наверное тут должно быть наоборот)

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/Main.java:
```diff
@@ -0,0 +1,78 @@
+     * @param botName Имя бота
+     * @param botToken Токен, для подключения к боту
+     */
+    public static void startTelegramBot(String botName, String botToken){
```
Почему `public`?

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/consoleBot/ConsoleReader.java:
```diff
@@ -15,6 +19,10 @@ public class ConsoleReader implements Reader {
+        RequestData requestData = new RequestData();
+        requestData.setComponent("text", new TextComponent(){{
+            set(text);
+        }});
```
Компоненту можно текст передавать в конструктор

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/Component.java:
```diff
@@ -0,0 +1,6 @@
+
+public abstract class Component {
+    abstract public Object get();
+    abstract public void set(Object value);
```
Обычно abstract пишут после модификатора доступа

TheDoctorTimeLord replied:
Ещё кажется, что Component можно сделать interface

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/ComponentsContainer.java:
```diff
@@ -0,0 +1,62 @@
+public class ComponentsContainer {
+    protected final Map<String, Component> dataComponents;
+
+    public ComponentsContainer (ComponentsContainer oldData){
```
Лучше параметры указывать слитно с именем, а вот перед фигурными скобками можно добавить пробел

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/ComponentsContainer.java:
```diff
@@ -0,0 +1,62 @@
+    public ComponentsContainer (){
+        dataComponents = new HashMap<>() {{
+            put("text", new TextComponent());
+        }};
```
Сейчас не рекомендуют делать `{{ ... }}`. Вроде как это медленнее и работает менее стабильно

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/ComponentsContainer.java:
```diff
@@ -0,0 +1,62 @@
+        Component component = dataComponents.get(name);
+        if (component == null)
+            return null;
+        return (T)component;
```
Можно сократить до `return (T)dataComponents.get(name);`

GlazProject committed:
Обновлены и добавлены тесты
Переведено хранение данных на массив из контейнеров

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/ComponentsContainer.java:
```diff
@@ -0,0 +1,43 @@
+     * @param component Экземпляр компонента с данными
+     */
+    public void setComponent(Component component){
+        dataComponents.add(component);
```
Теоретически, можно проверить, что таких компонентов больше нет, чтобы не допустить дублирования

TheDoctorTimeLord replied:
Можно, кстати, сначала попытаться удалить аналогичный компонент из `dataComponents`, а после этого добавлять новый. Либо же использовать `Set` вместо `List` и реализовать у каждого компонента методы `equals` и `hashCode`

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/DataWriter.java:
```diff
@@ -0,0 +1,26 @@
+            @NotNull ComponentsContainer data
+    ){
+        TextComponent textComponent = new TextComponent();
+        textComponent.setText(text);
```
Может просто текст в конструктор поместим?) Тогда потребность в этом классе отпадёт

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/dataContainer/components/TelegramChatIDComponent.java:
```diff
@@ -0,0 +1,29 @@
+    private Long id;
+
+    public TelegramChatIDComponent(){}
+    public TelegramChatIDComponent(Long chatID){
```
Давай пустую строку между конструкторами поставим

TheDoctorTimeLord replied:
Я предлагаю через один единственный конструктор с получением необходимых данных реализовать все компоненты. Тогда надобность в метода `set...` отпадёт

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/BaseHandlersFacade.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Список всех обработчиков (фасадов), которым даётся возможность обработать данные
+     */
+    protected final List<Handler> handlers = new LinkedList<>();
```
В Java лучше сводить к минимуму использование `LinkedList`. Он строится на идее связного списка, из-за его элементы занимают дополнительную память на ссылки, а также они становятся разбросанными по памяти. Лучше всего использовать `ArrayList` - он построен на массиве ссылок на объекты. За счёт непрерывного хранения ссылок в памяти получается буст в скорости. Если интересно - можем мне в лс написать, я расскажу подробнее

TheDoctorTimeLord replied:
А зачем нам `handlers` - `protected`? Мы можем принимать все обработчики в конструктор с использованием фишки `Handler... handler`

GlazProject replied:
По первому замечанию согласен, а по второму нет. Если мы его сделаем public - то это в принципе не правильно. Если заменить на private, то мы не сможем к нему обращаться из классов наследников. Поэтому остаётся только private

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/HandlerFacade.java:
```diff
@@ -0,0 +1,34 @@
+     * Задание списка всех обработчиков (фасадов), которым даётся возможность обработать данные
+     */
+    public HandlerFacade(){
+        super();
```
`super` без аргументов можно не указывать

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/HandlerFacade.java:
```diff
@@ -0,0 +1,34 @@
+     *                         которым нужно дополнительно давать возможность обработать данные
+     */
+    public HandlerFacade(Handler[] optionalHandlers){
+        super();
```
Аналогично предыдущему

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/HandlerFacade.java:
```diff
@@ -0,0 +1,34 @@
+     * @param optionalHandlers Массив экземпляров обработчиков (фасадов),
+     *                         которым нужно дополнительно давать возможность обработать данные
+     */
+    public HandlerFacade(Handler[] optionalHandlers){
```
Можно использовать небольшую фишку: `public HandlerFacade(Handler... optionalHandlers) {`. По сути, с точки зрения конструктора или метода optionalHandlers - тоже будут массивом. Зато при использовании мы можем написать `new HandlerFacade(element1, element2, element3)` и так далее.

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/HandlerFacade.java:
```diff
@@ -0,0 +1,34 @@
+     */
+    public HandlerFacade(Handler[] optionalHandlers){
+        super();
+        Collections.addAll(handlers, optionalHandlers);
+        handlers.add(new CommandFacade());
+        handlers.add(new TextFacade());
+    }
```
Я думаю, что лучше всего будет, если мы передадим это через `super` в конструктор нашему классу-родителю

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/handlers/commandHandlers/CommandFacade.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * Задание списка всех обработчиков, которым даётся возможность обработать данные
+     */
+    public CommandFacade(){
+        super();
+        handlers.add(new StartCommandHandler());
+        handlers.add(new UnknownCommandHandler());
+    }
```
Тут аналогично предыдущему

TheDoctorTimeLord commented on src/main/java/ru/telegramBot/gm/telegramBot/TelegramHandler.java:
```diff
@@ -0,0 +1,38 @@
+        if (data == null || data.getComponent(TelegramChatIDComponent.class) == null)
+            return null;
+
+        HandlerFacade handler = new HandlerFacade();
```
Давай его один раз создадим на уровне класса?

TheDoctorTimeLord reviewed (approved):
В целом я готов зачесть задачу. Предлагаю как-нибудь созвониться и вычистить лишние классы, которые остались. Думаю так будет лучше закончить работку этой задачи)

GlazProject committed:
Исправлены мелкие недочёты в обработчиках и контейнерах

GlazProject committed:
Составлено визуальное представление проекта

GlazProject committed:
Разбиение на пакеты
Рефакторинг кода
Добавлен DEVELOPMENT.md
Обновлён README.md

GlazProject committed:
Обновлён DEVELOPMENT.md
